Make Expr(:invoke) target be a CodeInstance, not MethodInstance

vtjnash opened the pull request:
This is a look at what is necessary and what it would look like for codegen to be using CodeInstance directly to get the API for a function call (specifically the rettype), rather than using a lookup call at that point to decide upon the API. It is based around the idea that eventually we would keep track of these anyways to form a graph of the inferred edge data, for use later in validation anyways (instead of attempting to invert the backedges graph in staticdata_utils.c), so we might as well use the same target type for the :invoke call representation also.

~~This does not depend upon https://github.com/JuliaLang/julia/pull/54894, but is much better with it. It did depend upon #54738, though~~ there still remains work to do to teach it how to select a correct alternative CodeInstance after deserializing. Also still to do: codegen also should be looking for equivalent objects, not just using the exact one given, per the definition of equivalency in https://hackmd.io/@vtjnash/codeinstances (and that is what precompile should be doing also).

Depends on ~~https://github.com/JuliaLang/julia/pull/56552~~, ~~https://github.com/JuliaLang/julia/pull/56551~~, ~~https://github.com/JuliaLang/julia/pull/56547~~, ~~https://github.com/JuliaLang/julia/pull/56598~~

vtjnash commented:
Unlike https://github.com/JuliaLang/julia/pull/52797, the runtime in this PR currently still treats a CodeInstance as simply an added ABI suggestion for a MethodInstance, but doesn't respect the owner field (nor did #52797 for that matter, which was implemented before the owner field). I don't think that is blocking for merging this, but it probably will need to be fixed soon.

vtjnash commented on base/compiler/ssair/inlining.jl:
```diff
@@ -1539,8 +1549,6 @@ function handle_modifyop!_call!(ir::IRCode, idx::Int, stmt::Expr, info::ModifyOp
     edge = info.edges[1]
```
@aviatesk is this the right API calls to use to extract the CodeInstance for a result? It is now the only place we call `compileable_specialization` with a MethodInstance instead of a CodeInstance, so it would be nice to get to simplify this callsite too

vtjnash replied:
okay, I thought other parts were using `getresult`, so I wasn't sure if that applied here

aviatesk commented:
There is this another place where `compileable_specialization(::MethodInstance, ...)` is used:  
https://github.com/JuliaLang/julia/blob/b10df4baf56241efd65237b36227e6adff8f21bf/Compiler/src/ssair/inlining.jl#L869  

I recall that most of these cases were due to inference failures, but simply doing `return nothing` here caused some regressions, if I remember correctly.

vtjnash commented on base/essentials.jl:
```diff
@@ -1042,6 +1042,7 @@ call obsolete versions of a function `f`.
     Prior to Julia 1.9, this function was not exported, and was called as `Base.invokelatest`.
 """
 function invokelatest(@nospecialize(f), @nospecialize args...; kwargs...)
+    @inline
```
Just noticing that codegen was doing dumb stuff here because of the nospecialize. It would generate a call to make a Tuple for the args so that it could call the apply function to unpack those again to call the builtin 🙄

vtjnash commented:
> simply doing return nothing here caused some regressions

I tried that and confirmed: it especially regresses the case where we found there was a get_compileable_sig different from sig, and that it was therefore wasteful to infer the sig, but in that case we still want to call compileable_specialization to find the get_compileable_sig target and use that as the invoke target. I updated the comment here to mention this.

Keno commented on src/interpreter.c:
```diff
@@ -135,8 +135,9 @@ static jl_value_t *do_invoke(jl_value_t **args, size_t nargs, interpreter_state
-    assert(jl_is_method_instance(meth));
+    jl_value_t *c = args[0];
+    assert(jl_is_code_instance(c) || jl_is_method_instance(c));
+    jl_method_instance_t *meth = jl_is_method_instance(c) ? (jl_method_instance_t*)c : ((jl_code_instance_t*)c)->def;
```
This should probably just invoke the codeinstance directly.

vtjnash replied:
Perhaps. We still have to support the legacy thing for a bit longer, at least while we solve a few last inference / optimizer bugs around this

vtjnash committed:
use CodeInstance for :invoke instead of MethodInstance, when possible

Make :invoke edges to CodeInstance (with rettype and edges) instead of
MethodInstance, for more accuracy (except when inference made a mistake).

Remove lookup parameter from codegen, since we no longer do any lookup.

vtjnash commented:
One stacked PR remaining (https://github.com/JuliaLang/julia/pull/56598), but this is already passing tests, so it can be merged once that is done

vtjnash commented:
This should be ready to go now, since it seems working, it would be nice to merge before it picks up spurious conflicts.

@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

Some improvements can be done later:
 - jitlayers/aotcompile/staticdata should be looking for equivalent objects instead of deferring that entirely to the runtime (as a jl_invoke)
 - conversely, interpreter should be trying to execute the object directly, instead of immediately looking for an equivalent one
 - non-nothing owner needs special handling implemented (currently it is still illegal to expect to put a different owner there), since codegen may become inconsistent and incorrect

Keno commented:
@vtjnash small conflicts with my worldage PR, so I rebased this for you, just FYI
